crypto/fipsmodule/ec/ec_test.cc

```diff
@@ -182,6 +182,259 @@ TEST(ECTest, Encoding) {
+  0x3f, 0x92, 0x72, 0x9b, 0x31, 0xc5, 0x5f, 0x7b
+};
+
+static const uint8_t kP384PublicKey_uncompressed[] = {
+  /* uncompressed */
+  0x04,
+  /* x-coordinate */
+  0xe4, 0xe7, 0x0e, 0x43, 0xc6, 0xd0, 0x43, 0x46, 0xdd, 0xd7, 0x62, 0xa6,
+  0x14, 0x17, 0x6d, 0x22, 0x78, 0xb0, 0x47, 0xc5, 0xec, 0x28, 0x64, 0x84,
+  0x65, 0xf2, 0xa3, 0x90, 0xf6, 0xdd, 0x6b, 0xba, 0x54, 0xb9, 0x0b, 0x1e,
+  0x62, 0xb3, 0x91, 0x85, 0xf8, 0xf3, 0x95, 0xf6, 0x65, 0x73, 0x6d, 0x1d,
+  /* y-coordinate */
+  0xf9, 0x62, 0xa2, 0x73, 0x6a, 0xce, 0x52, 0x56, 0x18, 0x15, 0xd5, 0x99,
+  0x53, 0xa0, 0x19, 0x1b, 0x1f, 0xb1, 0xf2, 0x88, 0xa4, 0x5f, 0x8e, 0x28,
+  0x3d, 0x40, 0xa5, 0xff, 0x0e, 0x83, 0x3f, 0xf3, 0x0b, 0xd6, 0x05, 0xb1,
+  0x0c, 0xf8, 0xc2, 0x6c, 0x57, 0x4d, 0x4c, 0x2f, 0x0d, 0xcd, 0xce, 0x21
+};
+
+static const uint8_t kP384PublicKey_compressed[] = {
+  0x03,
+  /* x-coordinate */
+  0xe4, 0xe7, 0x0e, 0x43, 0xc6, 0xd0, 0x43, 0x46, 0xdd, 0xd7, 0x62, 0xa6,
+  0x14, 0x17, 0x6d, 0x22, 0x78, 0xb0, 0x47, 0xc5, 0xec, 0x28, 0x64, 0x84,
+  0x65, 0xf2, 0xa3, 0x90, 0xf6, 0xdd, 0x6b, 0xba, 0x54, 0xb9, 0x0b, 0x1e,
+  0x62, 0xb3, 0x91, 0x85, 0xf8, 0xf3, 0x95, 0xf6, 0x65, 0x73, 0x6d, 0x1d
+};
+
+static const uint8_t kP521PublicKey_uncompressed[] = {
+  /* uncompressed */
+  0x04,
+  /* x-coordinate */
+  0x01, 0x03, 0x7e, 0x95, 0xff, 0x8e, 0x40, 0x31, 0xe0, 0xb0, 0x36, 0x1c,
+  0x58, 0xc0, 0x62, 0x61, 0x39, 0x56, 0xaa, 0x30, 0x77, 0x0c, 0xed, 0x17,
+  0x15, 0xed, 0x1b, 0x4d, 0x34, 0x29, 0x33, 0x0f, 0xac, 0x2f, 0xc5, 0xc9,
+  0x3a, 0x69, 0xf7, 0x98, 0x63, 0x3a, 0x15, 0x75, 0x5e, 0x2d, 0xb8, 0x65,
+  0x09, 0x87, 0xf5, 0x75, 0x85, 0xcd, 0xe3, 0x51, 0x6b, 0x6d, 0xd0, 0xfc,
+  0x9f, 0x5f, 0xb4, 0xf8, 0xe7, 0x7b,
+  /* y-coordinate */
+  0x01, 0x1b, 0xba, 0xcc, 0x17, 0x80, 0x56, 0x8b, 0x9b, 0x32, 0xd4, 0x82,
+  0x3f, 0x32, 0x9a, 0x46, 0xd8, 0x39, 0x39, 0xd1, 0x18, 0xcc, 0x97, 0x79,
+  0x8d, 0x5d, 0xfa, 0x08, 0xb4, 0x27, 0xd3, 0xae, 0xe4, 0x76, 0x4f, 0x46,
+  0x47, 0xf9, 0xf2, 0x4e, 0xcf, 0x0f, 0xee, 0x6d, 0x61, 0x9c, 0x79, 0x73,
+  0xa8, 0x55, 0x4a, 0xd5, 0x51, 0x13, 0x0d, 0x1e, 0x3f, 0x6c, 0x9d, 0x2e,
+  0xe3, 0xa2, 0xa8, 0x6f, 0xf5, 0xc3
+};
+
+static const uint8_t kP521PublicKey_compressed[] = {
+  0x03,
+  /* x-coordinate */
+  0x01, 0x03, 0x7e, 0x95, 0xff, 0x8e, 0x40, 0x31, 0xe0, 0xb0, 0x36, 0x1c,
+  0x58, 0xc0, 0x62, 0x61, 0x39, 0x56, 0xaa, 0x30, 0x77, 0x0c, 0xed, 0x17,
+  0x15, 0xed, 0x1b, 0x4d, 0x34, 0x29, 0x33, 0x0f, 0xac, 0x2f, 0xc5, 0xc9,
+  0x3a, 0x69, 0xf7, 0x98, 0x63, 0x3a, 0x15, 0x75, 0x5e, 0x2d, 0xb8, 0x65,
+  0x09, 0x87, 0xf5, 0x75, 0x85, 0xcd, 0xe3, 0x51, 0x6b, 0x6d, 0xd0, 0xfc,
+  0x9f, 0x5f, 0xb4, 0xf8, 0xe7, 0x7b
+};
+
+struct ECPublicKeyTestInput {
+  const uint8_t *input_key;
+  size_t input_key_len;
+  point_conversion_form_t encode_conv_form;
+  const uint8_t *expected_output_key;
+  size_t expected_output_key_len;
+  int nid;
+} kDecodeAndEncodeInputs[] = {
+    /* Test 1: decode uncompressed |EC_KEY|, and then encode with the same |conv_form|. */
```

Comment:
I like these tests.

Should we also test behaviour for unhappy cases? e.g. what is the expected behaviour if we set the conversion form to uncompressed but input a compressed point?

It should probably return an error when the decode function reads the first two encoding bytes, detecting a mismatch.

Also, what happens if we have a point that is compressed but with a wrong two first encoding bytes? e.g. `kP521PublicKey_compressed` starts with `0x02`...

Or... maybe all this is covered by some other test suite?

Reply:
> what is the expected behaviour if we set the conversion form to uncompressed but input a compressed point?

This should be happy case and included in this test file.

> Also, what happens if we have a point that is compressed but with a wrong two first encoding bytes? e.g. kP521PublicKey_compressed starts with 0x02...

I can add this type unhappy cases. If you have more suggestions, feel free to publish new comments.

> Or... maybe all this is covered by some other test suite?

I would say no. AFAIK, before this PR, one method (encode or decode, forgot which one) has never processed data relating to `COMPRESSED` point. I added some print statements to the methods but only saw  output relatign to `UNCOMPRESSED` points.

Reply:
Ahh so the curve point input decides the conversion form. But for encode you need to specify the conversion form.